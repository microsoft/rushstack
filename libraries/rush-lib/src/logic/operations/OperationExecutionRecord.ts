// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import {
  type ITerminal,
  type ITerminalProvider,
  DiscardStdoutTransform,
  SplitterTransform,
  StderrLineTransform,
  StdioSummarizer,
  TextRewriterTransform,
  Terminal,
  type TerminalWritable
} from '@rushstack/terminal';
import { InternalError, NewlineKind } from '@rushstack/node-core-library';
import { CollatedTerminal, type CollatedWriter, type StreamCollator } from '@rushstack/stream-collator';

import { OperationStatus, TERMINAL_STATUSES } from './OperationStatus';
import type { IOperationRunner, IOperationRunnerContext } from './IOperationRunner';
import type { Operation } from './Operation';
import { Stopwatch } from '../../utilities/Stopwatch';
import { OperationMetadataManager } from './OperationMetadataManager';
import type { IPhase } from '../../api/CommandLineConfiguration';
import type { RushConfigurationProject } from '../../api/RushConfigurationProject';
import { CollatedTerminalProvider } from '../../utilities/CollatedTerminalProvider';
import {
  getProjectLogFilePaths,
  type ILogFilePaths,
  initializeProjectLogFilesAsync
} from './ProjectLogWritable';
import type { IOperationExecutionResult } from './IOperationExecutionResult';

export interface IOperationExecutionRecordContext {
  streamCollator: StreamCollator;
  onOperationStatusChanged?: (record: OperationExecutionRecord) => void;

  debugMode: boolean;
  quietMode: boolean;
}

/**
 * Internal class representing everything about executing an operation
 *
 * @internal
 */
export class OperationExecutionRecord implements IOperationRunnerContext, IOperationExecutionResult {
  /**
   * The associated operation.
   */
  public readonly operation: Operation;

  /**
   * The error which occurred while executing this operation, this is stored in case we need
   * it later (for example to re-print errors at end of execution).
   */
  public error: Error | undefined = undefined;

  /**
   * This number represents how far away this Operation is from the furthest "root" operation (i.e.
   * an operation with no consumers). This helps us to calculate the critical path (i.e. the
   * longest chain of projects which must be executed in order, thereby limiting execution speed
   * of the entire operation tree.
   *
   * This number is calculated via a memoized depth-first search, and when choosing the next
   * operation to execute, the operation with the highest criticalPathLength is chosen.
   *
   * Example:
   * ```
   *        (0) A
   *             \
   *          (1) B     C (0)         (applications)
   *               \   /|\
   *                \ / | \
   *             (2) D  |  X (1)      (utilities)
   *                    | / \
   *                    |/   \
   *                (2) Y     Z (2)   (other utilities)
   *
   * All roots (A & C) have a criticalPathLength of 0.
   * B has a score of 1, since A depends on it.
   * D has a score of 2, since we look at the longest chain (e.g D->B->A is longer than D->C)
   * X has a score of 1, since the only package which depends on it is A
   * Z has a score of 2, since only X depends on it, and X has a score of 1
   * Y has a score of 2, since the chain Y->X->C is longer than Y->C
   * ```
   *
   * The algorithm is implemented in AsyncOperationQueue.ts as calculateCriticalPathLength()
   */
  public criticalPathLength: number | undefined = undefined;

  /**
   * The set of operations that must complete before this operation executes.
   */
  public readonly dependencies: Set<OperationExecutionRecord> = new Set();
  /**
   * The set of operations that depend on this operation.
   */
  public readonly consumers: Set<OperationExecutionRecord> = new Set();

  public readonly stopwatch: Stopwatch = new Stopwatch();
  public readonly stdioSummarizer: StdioSummarizer = new StdioSummarizer({
    // Allow writing to this object after transforms have been closed. We clean it up manually in a finally block.
    preventAutoclose: true
  });

  public readonly runner: IOperationRunner;
  public readonly associatedPhase: IPhase | undefined;
  public readonly associatedProject: RushConfigurationProject | undefined;
  public readonly _operationMetadataManager: OperationMetadataManager | undefined;

  public logFilePaths: ILogFilePaths | undefined;

  private readonly _context: IOperationExecutionRecordContext;

  private _collatedWriter: CollatedWriter | undefined = undefined;
  private _status: OperationStatus;

  public constructor(operation: Operation, context: IOperationExecutionRecordContext) {
    const { runner, associatedPhase, associatedProject } = operation;

    if (!runner) {
      throw new InternalError(
        `Operation for phase '${associatedPhase?.name}' and project '${associatedProject?.packageName}' has no runner.`
      );
    }

    this.operation = operation;
    this.runner = runner;
    this.associatedPhase = associatedPhase;
    this.associatedProject = associatedProject;
    this.logFilePaths = undefined;

    this._operationMetadataManager =
      associatedPhase && associatedProject
        ? new OperationMetadataManager({
            phase: associatedPhase,
            rushProject: associatedProject,
            operation
          })
        : undefined;

    this._context = context;
    this._status = operation.dependencies.size > 0 ? OperationStatus.Waiting : OperationStatus.Ready;
  }

  public get name(): string {
    return this.runner.name;
  }

  public get weight(): number {
    return this.operation.weight;
  }

  public get debugMode(): boolean {
    return this._context.debugMode;
  }

  public get quietMode(): boolean {
    return this._context.quietMode;
  }

  public get collatedWriter(): CollatedWriter {
    // Lazy instantiate because the registerTask() call affects display ordering
    if (!this._collatedWriter) {
      this._collatedWriter = this._context.streamCollator.registerTask(this.name);
    }
    return this._collatedWriter;
  }

  public get nonCachedDurationMs(): number | undefined {
    // Lazy calculated because the state file is created/restored later on
    return this._operationMetadataManager?.stateFile.state?.nonCachedDurationMs;
  }

  public get cobuildRunnerId(): string | undefined {
    // Lazy calculated because the state file is created/restored later on
    return this._operationMetadataManager?.stateFile.state?.cobuildRunnerId;
  }

  public get metadataFolderPath(): string | undefined {
    return this._operationMetadataManager?.metadataFolderPath;
  }

  public get isTerminal(): boolean {
    return TERMINAL_STATUSES.has(this.status);
  }

  /**
   * The current execution status of an operation. Operations start in the 'ready' state,
   * but can be 'blocked' if an upstream operation failed. It is 'executing' when
   * the operation is executing. Once execution is complete, it is either 'success' or
   * 'failure'.
   */
  public get status(): OperationStatus {
    return this._status;
  }
  public set status(newStatus: OperationStatus) {
    if (newStatus === this._status) {
      return;
    }
    this._status = newStatus;
    this._context.onOperationStatusChanged?.(this);
  }

  public get silent(): boolean {
    return !this.operation.enabled || this.runner.silent;
  }

  /**
   * {@inheritdoc IOperationRunnerContext.runWithTerminalAsync}
   */
  public async runWithTerminalAsync<T>(
    callback: (terminal: ITerminal, terminalProvider: ITerminalProvider) => Promise<T>,
    options: {
      createLogFile: boolean;
      logFileSuffix: string;
    }
  ): Promise<T> {
    const { associatedPhase, associatedProject, stdioSummarizer } = this;
    const { createLogFile, logFileSuffix = '' } = options;

    const logFilePaths: ILogFilePaths | undefined =
      createLogFile && associatedProject && associatedPhase && this._operationMetadataManager
        ? getProjectLogFilePaths({
            project: associatedProject,
            logFilenameIdentifier: `${this._operationMetadataManager.logFilenameIdentifier}${logFileSuffix}`
          })
        : undefined;
    this.logFilePaths = logFilePaths;

    const projectLogWritable: TerminalWritable | undefined = logFilePaths
      ? await initializeProjectLogFilesAsync({
          logFilePaths,
          enableChunkedOutput: true
        })
      : undefined;

    try {
      //#region OPERATION LOGGING
      // TERMINAL PIPELINE:
      //
      //                             +--> quietModeTransform? --> collatedWriter
      //                             |
      // normalizeNewlineTransform --1--> stderrLineTransform --2--> projectLogWritable
      //                                                        |
      //                                                        +--> stdioSummarizer
      const destination: TerminalWritable = projectLogWritable
        ? new SplitterTransform({
            destinations: [projectLogWritable, stdioSummarizer]
          })
        : stdioSummarizer;

      const stderrLineTransform: StderrLineTransform = new StderrLineTransform({
        destination,
        newlineKind: NewlineKind.Lf // for StdioSummarizer
      });

      const splitterTransform1: SplitterTransform = new SplitterTransform({
        destinations: [
          this.quietMode
            ? new DiscardStdoutTransform({ destination: this.collatedWriter })
            : this.collatedWriter,
          stderrLineTransform
        ]
      });

      const normalizeNewlineTransform: TextRewriterTransform = new TextRewriterTransform({
        destination: splitterTransform1,
        normalizeNewlines: NewlineKind.Lf,
        ensureNewlineAtEnd: true
      });

      const collatedTerminal: CollatedTerminal = new CollatedTerminal(normalizeNewlineTransform);
      const terminalProvider: CollatedTerminalProvider = new CollatedTerminalProvider(collatedTerminal, {
        debugEnabled: this.debugMode
      });
      const terminal: Terminal = new Terminal(terminalProvider);
      //#endregion

      const result: T = await callback(terminal, terminalProvider);

      normalizeNewlineTransform.close();

      // If the pipeline is wired up correctly, then closing normalizeNewlineTransform should
      // have closed projectLogWritable.
      if (projectLogWritable?.isOpen) {
        throw new InternalError('The output file handle was not closed');
      }

      return result;
    } finally {
      projectLogWritable?.close();
    }
  }

  public async executeAsync({
    onStart,
    onResult
  }: {
    onStart: (record: OperationExecutionRecord) => Promise<OperationStatus | undefined>;
    onResult: (record: OperationExecutionRecord) => Promise<void>;
  }): Promise<void> {
    if (!this.isTerminal) {
      this.stopwatch.reset();
    }
    this.stopwatch.start();
    this.status = OperationStatus.Executing;

    try {
      const earlyReturnStatus: OperationStatus | undefined = await onStart(this);
      // When the operation status returns by the hook, bypass the runner execution.
      if (earlyReturnStatus) {
        this.status = earlyReturnStatus;
      } else {
        // If the operation is disabled, skip the runner and directly mark as Skipped.
        // However, if the operation is a NoOp, return NoOp so that cache entries can still be written.
        this.status = this.operation.enabled
          ? await this.runner.executeAsync(this)
          : this.runner.isNoOp
            ? OperationStatus.NoOp
            : OperationStatus.Skipped;
      }
      // Delegate global state reporting
      await onResult(this);
    } catch (error) {
      this.status = OperationStatus.Failure;
      this.error = error;
      // Delegate global state reporting
      await onResult(this);
    } finally {
      if (this.isTerminal) {
        this._collatedWriter?.close();
        this.stdioSummarizer.close();
        this.stopwatch.stop();
      }
    }
  }
}
