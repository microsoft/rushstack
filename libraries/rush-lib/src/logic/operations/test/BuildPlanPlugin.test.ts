// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
// See LICENSE in the project root for license information.

import { MockWritable, StringBufferTerminalProvider, Terminal } from '@rushstack/terminal';
import { JsonFile } from '@rushstack/node-core-library';
import { StreamCollator } from '@rushstack/stream-collator';
import { BuildPlanPlugin } from '../BuildPlanPlugin';
import {
  type ICreateOperationsContext,
  type IExecuteOperationsContext,
  PhasedCommandHooks
} from '../../../pluginFramework/PhasedCommandHooks';
import type { Operation } from '../Operation';
import { RushConfiguration } from '../../../api/RushConfiguration';
import {
  CommandLineConfiguration,
  type IPhase,
  type IPhasedCommandConfig
} from '../../../api/CommandLineConfiguration';
import { OperationExecutionRecord } from '../OperationExecutionRecord';
import { PhasedOperationPlugin } from '../PhasedOperationPlugin';
import type { RushConfigurationProject } from '../../../api/RushConfigurationProject';
import { RushConstants } from '../../RushConstants';
import { MockOperationRunner } from './MockOperationRunner';
import path from 'path';
import type { ICommandLineJson } from '../../../api/CommandLineJson';
import type { IInputsSnapshot } from '../../incremental/InputsSnapshot';

describe(BuildPlanPlugin.name, () => {
  const rushJsonFile: string = path.resolve(__dirname, `../../test/workspaceRepo/rush.json`);
  const commandLineJsonFile: string = path.resolve(
    __dirname,
    `../../test/workspaceRepo/common/config/rush/command-line.json`
  );
  let rushConfiguration!: RushConfiguration;
  let commandLineConfiguration!: CommandLineConfiguration;
  let stringBufferTerminalProvider!: StringBufferTerminalProvider;
  let terminal!: Terminal;
  const mockStreamWritable: MockWritable = new MockWritable();
  const streamCollator = new StreamCollator({
    destination: mockStreamWritable
  });
  beforeEach(() => {
    stringBufferTerminalProvider = new StringBufferTerminalProvider();
    terminal = new Terminal(stringBufferTerminalProvider);
    mockStreamWritable.reset();
    rushConfiguration = RushConfiguration.loadFromConfigurationFile(rushJsonFile);
    const commandLineJson: ICommandLineJson = JsonFile.load(commandLineJsonFile);

    commandLineConfiguration = new CommandLineConfiguration(commandLineJson);
  });

  function createMockRunner(operations: Set<Operation>, context: ICreateOperationsContext): Set<Operation> {
    for (const operation of operations) {
      const { associatedPhase, associatedProject } = operation;

      if (associatedPhase && associatedProject && !operation.runner) {
        const name: string = `${associatedProject.packageName} (${associatedPhase.name.slice(
          RushConstants.phaseNamePrefix.length
        )})`;

        operation.runner = new MockOperationRunner(name, undefined, undefined, false);
      }
    }

    return operations;
  }

  async function testCreateOperationsAsync(
    phaseSelection: Set<IPhase>,
    projectSelection: Set<RushConfigurationProject>,
    changedProjects: Set<RushConfigurationProject>
  ): Promise<Set<Operation>> {
    const hooks: PhasedCommandHooks = new PhasedCommandHooks();
    // Apply the plugin being tested
    new PhasedOperationPlugin().apply(hooks);
    // Add mock runners for included operations.
    hooks.createOperations.tap('MockOperationRunnerPlugin', createMockRunner);

    const context: Pick<
      ICreateOperationsContext,
      | 'phaseOriginal'
      | 'phaseSelection'
      | 'projectSelection'
      | 'projectsInUnknownState'
      | 'projectConfigurations'
    > = {
      phaseOriginal: phaseSelection,
      phaseSelection,
      projectSelection,
      projectsInUnknownState: changedProjects,
      projectConfigurations: new Map()
    };
    const operations: Set<Operation> = await hooks.createOperations.promise(
      new Set(),
      context as ICreateOperationsContext
    );

    return operations;
  }

  describe('build plan debugging', () => {
    it('should generate a build plan', async () => {
      const hooks: PhasedCommandHooks = new PhasedCommandHooks();

      new BuildPlanPlugin(terminal).apply(hooks);
      const inputsSnapshot: Pick<IInputsSnapshot, 'getTrackedFileHashesForOperation'> = {
        getTrackedFileHashesForOperation() {
          return new Map();
        }
      };
      const context: Pick<IExecuteOperationsContext, 'inputsSnapshot' | 'projectConfigurations'> = {
        inputsSnapshot: inputsSnapshot as unknown as IInputsSnapshot,
        projectConfigurations: new Map()
      };
      const buildCommand: IPhasedCommandConfig = commandLineConfiguration.commands.get(
        'build'
      )! as IPhasedCommandConfig;

      const operationMap = new Map();

      const operations = await testCreateOperationsAsync(
        buildCommand.phases,
        new Set(rushConfiguration.projects),
        new Set(rushConfiguration.projects)
      );
      operations.forEach((operation) => {
        operationMap.set(
          operation,
          new OperationExecutionRecord(operation, { debugMode: false, quietMode: true, streamCollator })
        );
      });

      await hooks.beforeExecuteOperations.promise(operationMap, context as IExecuteOperationsContext);

      expect(stringBufferTerminalProvider.getOutput({ normalizeSpecialCharacters: false })).toMatchSnapshot();
    });
  });
});
