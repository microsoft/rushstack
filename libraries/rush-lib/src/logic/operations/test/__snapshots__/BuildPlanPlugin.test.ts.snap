// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BuildPlanPlugin build plan debugging should generate a build plan 1`] = `
"Build Plan Depth (deepest dependency tree): 5
Build Plan Width (maximum parallelism): 38
Number of Nodes per Depth: 22, 38, 33, 11, 1
Plan @ Depth 0 has 22 nodes and 0 dependents:
- a (upstream-3)
- a (no-deps)
- a (upstream-1)
- a (upstream-1-self-upstream)
- a (upstream-2)
- b (no-deps)
- c (no-deps)
- d (no-deps)
- e (no-deps)
- f (no-deps)
- g (no-deps)
- h (no-deps)
- i (upstream-3)
- i (no-deps)
- i (upstream-1)
- i (upstream-1-self-upstream)
- i (upstream-2)
- j (upstream-3)
- j (no-deps)
- j (upstream-1)
- j (upstream-1-self-upstream)
- j (upstream-2)
Plan @ Depth 1 has 38 nodes and 22 dependents:
- a (complex)
- a (upstream-self)
- b (upstream-1)
- f (upstream-1)
- g (upstream-1)
- h (upstream-1)
- b (upstream-2)
- f (upstream-2)
- g (upstream-2)
- h (upstream-2)
- a (upstream-1-self)
- b (complex)
- f (complex)
- g (complex)
- h (complex)
- b (upstream-3)
- f (upstream-3)
- g (upstream-3)
- h (upstream-3)
- a (upstream-2-self)
- b (upstream-self)
- c (upstream-1)
- d (upstream-1)
- c (upstream-self)
- e (upstream-1)
- d (upstream-self)
- e (upstream-self)
- f (upstream-self)
- g (upstream-self)
- h (upstream-self)
- i (complex)
- i (upstream-self)
- i (upstream-1-self)
- i (upstream-2-self)
- j (complex)
- j (upstream-self)
- j (upstream-1-self)
- j (upstream-2-self)
Plan @ Depth 2 has 33 nodes and 60 dependents:
- b (upstream-self)
- f (upstream-self)
- h (upstream-self)
- g (upstream-self)
- c (upstream-2)
- d (upstream-2)
- b (upstream-1-self)
- f (upstream-1-self)
- g (upstream-1-self)
- f (upstream-2)
- h (upstream-1-self)
- c (upstream-3)
- d (upstream-3)
- b (upstream-2-self)
- f (upstream-2-self)
- g (upstream-2-self)
- f (upstream-3)
- h (upstream-2-self)
- b (upstream-1-self-upstream)
- f (upstream-1-self-upstream)
- g (upstream-1-self-upstream)
- h (upstream-1-self-upstream)
- b (complex)
- f (complex)
- g (complex)
- h (complex)
- c (upstream-self)
- d (upstream-self)
- e (upstream-2)
- c (upstream-1-self)
- d (upstream-1-self)
- e (upstream-self)
- e (upstream-1-self)
Plan @ Depth 3 has 11 nodes and 93 dependents:
- e (upstream-3)
- c (upstream-2-self)
- d (upstream-2-self)
- c (upstream-1-self-upstream)
- d (upstream-1-self-upstream)
- f (upstream-1-self-upstream)
- c (complex)
- d (complex)
- f (complex)
- e (upstream-2-self)
- e (upstream-1-self-upstream)
Plan @ Depth 4 has 1 nodes and 104 dependents:
- e (complex)
##################################################
                  a (complex): (0)
               a (upstream-3): (0)
                  a (no-deps): (1)
               a (upstream-1): (1)
 a (upstream-1-self-upstream): (1)
               a (upstream-2): (1)
                  b (no-deps): (1)
                  c (no-deps): (1)
                  d (no-deps): (1)
                  e (no-deps): (1)
                  f (no-deps): (1)
                  g (no-deps): (1)
                  h (no-deps): (1)
                  i (complex): (2)
               i (upstream-3): (2)
                  i (no-deps): (3)
               i (upstream-1): (4)
 i (upstream-1-self-upstream): (5)
               i (upstream-2): (6)
                  j (complex): (7)
               j (upstream-3): (7)
                  j (no-deps): (8)
               j (upstream-1): (9)
 j (upstream-1-self-upstream): (10)
               j (upstream-2): (11)
          a (upstream-1-self): -(1)
          a (upstream-2-self): -(1)
            a (upstream-self): -(1)
               b (upstream-1): -(1)
               b (upstream-2): -(1)
               b (upstream-3): -(1)
               c (upstream-1): -(1)
               d (upstream-1): -(1)
               e (upstream-1): -(1)
               f (upstream-1): -(1)
               f (upstream-2): -(1)
               f (upstream-3): -(1)
               g (upstream-1): -(1)
               g (upstream-2): -(1)
               g (upstream-3): -(1)
               h (upstream-1): -(1)
               h (upstream-2): -(1)
               h (upstream-3): -(1)
            i (upstream-self): -(3)
          i (upstream-1-self): -(4)
          i (upstream-2-self): -(6)
            j (upstream-self): -(8)
          j (upstream-1-self): -(9)
          j (upstream-2-self): -(11)
                  b (complex): --(1)
          b (upstream-1-self): --(1)
 b (upstream-1-self-upstream): --(1)
          b (upstream-2-self): --(1)
            b (upstream-self): --(1)
          c (upstream-1-self): --(1)
               c (upstream-2): --(1)
               c (upstream-3): --(1)
          d (upstream-1-self): --(1)
               d (upstream-2): --(1)
               d (upstream-3): --(1)
          e (upstream-1-self): --(1)
               e (upstream-2): --(1)
                  f (complex): --(1)
          f (upstream-1-self): --(1)
 f (upstream-1-self-upstream): --(1)
          f (upstream-2-self): --(1)
            f (upstream-self): --(1)
                  g (complex): --(1)
          g (upstream-1-self): --(1)
 g (upstream-1-self-upstream): --(1)
          g (upstream-2-self): --(1)
            g (upstream-self): --(1)
                  h (complex): --(1)
          h (upstream-1-self): --(1)
 h (upstream-1-self-upstream): --(1)
          h (upstream-2-self): --(1)
            h (upstream-self): --(1)
                  c (complex): ---(1)
 c (upstream-1-self-upstream): ---(1)
          c (upstream-2-self): ---(1)
            c (upstream-self): ---(1)
                  d (complex): ---(1)
 d (upstream-1-self-upstream): ---(1)
          d (upstream-2-self): ---(1)
            d (upstream-self): ---(1)
 e (upstream-1-self-upstream): ---(1)
          e (upstream-2-self): ---(1)
               e (upstream-3): ---(1)
                  e (complex): ----(1)
            e (upstream-self): ----(1)
##################################################
Cluster 0:
- Dependencies: none
- Clustered by: 
  - (a (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: a (complex), a (upstream-3)
--------------------------------------------------
Cluster 1:
- Dependencies: none
- Clustered by: 
  - (a (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (a (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (b (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (d (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (d (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (d (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (d (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (e (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (e (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (e (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (e (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (c (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (f (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-1-self-upstream)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-2-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (f (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-1-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (f (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (f (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-self)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (g (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (g (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (g (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (g (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
  - (h (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: a (no-deps), a (upstream-1), a (upstream-1-self), a (upstream-1-self-upstream), a (upstream-2), a (upstream-2-self), a (upstream-self), b (complex), b (no-deps), b (upstream-1), b (upstream-1-self), b (upstream-1-self-upstream), b (upstream-2), b (upstream-2-self), b (upstream-3), b (upstream-self), c (complex), c (no-deps), c (upstream-1), c (upstream-1-self), c (upstream-1-self-upstream), c (upstream-2), c (upstream-2-self), c (upstream-3), c (upstream-self), d (complex), d (no-deps), d (upstream-1), d (upstream-1-self), d (upstream-1-self-upstream), d (upstream-2), d (upstream-2-self), d (upstream-3), d (upstream-self), e (complex), e (no-deps), e (upstream-1), e (upstream-1-self), e (upstream-1-self-upstream), e (upstream-2), e (upstream-2-self), e (upstream-3), e (upstream-self), f (complex), f (no-deps), f (upstream-1), f (upstream-1-self), f (upstream-1-self-upstream), f (upstream-2), f (upstream-2-self), f (upstream-3), f (upstream-self), g (complex), g (no-deps), g (upstream-1), g (upstream-1-self), g (upstream-1-self-upstream), g (upstream-2), g (upstream-2-self), g (upstream-3), g (upstream-self), h (complex), h (no-deps), h (upstream-1), h (upstream-1-self), h (upstream-1-self-upstream), h (upstream-2), h (upstream-2-self), h (upstream-3), h (upstream-self)
--------------------------------------------------
Cluster 2:
- Dependencies: none
- Clustered by: 
  - (i (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: i (complex), i (upstream-3)
--------------------------------------------------
Cluster 3:
- Dependencies: none
- Clustered by: 
  - (i (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: i (no-deps), i (upstream-self)
--------------------------------------------------
Cluster 4:
- Dependencies: none
- Clustered by: 
  - (i (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: i (upstream-1), i (upstream-1-self)
--------------------------------------------------
Cluster 5:
- Dependencies: none
- Operations: i (upstream-1-self-upstream)
--------------------------------------------------
Cluster 6:
- Dependencies: none
- Clustered by: 
  - (i (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: i (upstream-2), i (upstream-2-self)
--------------------------------------------------
Cluster 7:
- Dependencies: none
- Clustered by: 
  - (j (upstream-3)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: j (complex), j (upstream-3)
--------------------------------------------------
Cluster 8:
- Dependencies: none
- Clustered by: 
  - (j (no-deps)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: j (no-deps), j (upstream-self)
--------------------------------------------------
Cluster 9:
- Dependencies: none
- Clustered by: 
  - (j (upstream-1)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: j (upstream-1), j (upstream-1-self)
--------------------------------------------------
Cluster 10:
- Dependencies: none
- Operations: j (upstream-1-self-upstream)
--------------------------------------------------
Cluster 11:
- Dependencies: none
- Clustered by: 
  - (j (upstream-2)) \\"Project does not have a rush-project.json configuration file, or one provided by a rig, so it does not support caching.\\"
- Operations: j (upstream-2), j (upstream-2-self)
--------------------------------------------------
##################################################
"
`;
